core/lib/spree/event/configuration.rb

```diff
@@ -14,12 +18,64 @@ def autoload_subscribers
       end
 
       def adapter
-        @adapter ||= Spree::Event::Adapters::ActiveSupportNotifications
+        @adapter ||= Spree::Core::VersionedValue.new(
+          Spree::Event::Adapters::ActiveSupportNotifications,
+          "4.0.0.alpha" => Spree::Event::Adapters::EventBus.new
+        ).call.tap do |value|
+          deprecate_if_legacy_adapter(value)
+        end
       end
 
+      # Only used by {Spree::Event::Adapters::ActiveSupportNotifications}.
       def suffix
         @suffix ||= '.spree'
       end
+
+      private
+
+      def deprecate_if_legacy_adapter(adapter)
+        Spree::Deprecation.warn <<~MSG if adapter == Spree::Event::Adapters::ActiveSupportNotifications && !ENV['CI']
```

Comment:
Why we need `!ENV['CI']`?

Reply:
The current version of Solidus is still using the legacy adapter, even though it's warning about its deprecation. CircleCI is currently configured to raise when any warning is displayed:

https://github.com/solidusio/solidus/blob/7118543095e0db0887f3b851e1107a1afe5b6aba/.circleci/config.yml#L11

The `CI` environment variable is set on CircleCI, making it possible to have a green build.

However, the warnings are indeed displayed when running `bin/rspec` locally. We could create a custom `ENV` variable in the `spec_helper` files and check it instead of `CI`, but I'm not a fan of this solution as it's too easy for that variable definitions to say there once they're no longer needed. Remember that we can't just ignore the warning when `RAILS_ENV` is `test` as we do want it to be displayed on users' application test suite. Do you think we have a better alternative?

Reply:
> Do you think we have a better alternative?

I think we could set up the dummy app to test against the new adapter. This means that the deprecated code won't be tested by default, which is not ideal because people may still use it. We could build a CI matrix (or at least one job) that tests Solidus with the legacy adapter, or we can just create some smoke tests for it, changing the adapter runtime before running those specs. Does it make sense?

Reply:
Yeah, I think that's the best option we have. I updated the code so that we're using the new adapter on the dummy application. As for the smoke tests, we can use the same that we had before, those on `event_spec.rb` that reach the legacy adapter.